test/unit/util/bitvector_black.h

```diff
@@ -26,30 +26,147 @@ class BitVectorBlack : public CxxTest::TestSuite {
 
 
 public:
+ void setUp()
+ {
+   zero = BitVector(4);
+   one = zero.setBit(0);
+   two = BitVector("0010", 2);
+   negOne = BitVector(4, Integer(-1));
+ }
 
-  void testStringConstructor() {
-    BitVector b1("0101",2);
-    TS_ASSERT_EQUALS( 4u, b1.getSize() );
-    TS_ASSERT_EQUALS( "0101", b1.toString() );
-    TS_ASSERT_EQUALS( "5", b1.toString(10) );
-    TS_ASSERT_EQUALS( "5", b1.toString(16) );
-
-    BitVector b2("000001", 2);
-    TS_ASSERT_EQUALS( 6u, b2.getSize() );
-    TS_ASSERT_EQUALS( "000001", b2.toString() );
-    TS_ASSERT_EQUALS( "1", b2.toString(10) );
-    TS_ASSERT_EQUALS( "1", b2.toString(16) );
-
-    BitVector b3("7f", 16);
-    TS_ASSERT_EQUALS( 8u, b3.getSize() );
-    TS_ASSERT_EQUALS( "01111111", b3.toString() );
-    TS_ASSERT_EQUALS( "127", b3.toString(10) );
-    TS_ASSERT_EQUALS( "7f", b3.toString(16) );
-
-    BitVector b4("01a", 16);
-    TS_ASSERT_EQUALS( 12u, b4.getSize() );
-    TS_ASSERT_EQUALS( "000000011010", b4.toString() );
-    TS_ASSERT_EQUALS( "26", b4.toString(10) );
-    TS_ASSERT_EQUALS( "1a", b4.toString(16) );
+ void testStringConstructor()
+ {
+   BitVector b1("0101", 2);
+   TS_ASSERT_EQUALS(4u, b1.getSize());
+   TS_ASSERT_EQUALS("0101", b1.toString());
+   TS_ASSERT_EQUALS("5", b1.toString(10));
+   TS_ASSERT_EQUALS("5", b1.toString(16));
+
+   BitVector b2("000001", 2);
+   TS_ASSERT_EQUALS(6u, b2.getSize());
+   TS_ASSERT_EQUALS("000001", b2.toString());
+   TS_ASSERT_EQUALS("1", b2.toString(10));
+   TS_ASSERT_EQUALS("1", b2.toString(16));
+
+   BitVector b3("7f", 16);
+   TS_ASSERT_EQUALS(8u, b3.getSize());
+   TS_ASSERT_EQUALS("01111111", b3.toString());
+   TS_ASSERT_EQUALS("127", b3.toString(10));
+   TS_ASSERT_EQUALS("7f", b3.toString(16));
+
+   BitVector b4("01a", 16);
+   TS_ASSERT_EQUALS(12u, b4.getSize());
+   TS_ASSERT_EQUALS("000000011010", b4.toString());
+   TS_ASSERT_EQUALS("26", b4.toString(10));
+   TS_ASSERT_EQUALS("1a", b4.toString(16));
+  }
+
+  void testConversions()
+  {
+    TS_ASSERT_EQUALS(two.toSignedInteger(), Integer(2));
+    TS_ASSERT_EQUALS("1111", negOne.toString());
+    TS_ASSERT_EQUALS(negOne.getValue(), Integer(15));
+    TS_ASSERT_EQUALS(negOne.getSize(), 4);
+    TS_ASSERT_EQUALS(negOne.toInteger(), Integer(15));
+    TS_ASSERT_EQUALS(negOne.toSignedInteger(), Integer(-1));
+
+    TS_ASSERT_EQUALS(zero.hash(), (one - one).hash());
+    TS_ASSERT_DIFFERS(negOne.hash(), zero.hash());
+  }
+
+  void testSetGetBit()
+  {
+    TS_ASSERT_EQUALS(one.setBit(1).setBit(2).setBit(3), negOne);
+
+    TS_ASSERT(negOne.isBitSet(3));
+    TS_ASSERT(!two.isBitSet(3));
+
+    TS_ASSERT_EQUALS(one.getValue(), Integer(1));
+    TS_ASSERT_EQUALS(zero.isPow2(), 0);
+    TS_ASSERT_EQUALS(one.isPow2(), 1);
+    TS_ASSERT_EQUALS(two.isPow2(), 2);
+  }
+
+  void testConcatExtract()
+  {
+    BitVector b = one.concat(zero);
+    TS_ASSERT_EQUALS(b.toString(), "00010000");
+    TS_ASSERT_EQUALS(b.extract(7, 4), one);
+    TS_ASSERT_EQUALS(b.concat(BitVector()), b);
```

Comment:
I'm missing error tests here, e.g., when `high` and `low` of extract do not match specifications:
`high < low`
`high >= d_size`
What happens if you do a `BitVector().extract(2, 1)` ?

Reply:
Added the error tests and error checks in BitVector::extract.